naming-conventions.md

```diff
@@ -0,0 +1,106 @@
+```
+
+`s` is for state and `e` is for environment
+```haskell
+put :: MonadState s m => s -> m ()
+ask :: MonadReader e m => m e
+```
+
+These are conventions only, but picking something unusual for a common typeclass might surprise other readers.
+
+### Concrete Types
+
+Avoid using single-letter names or abbreviations for values with a concrete type, especially if that type is domain-specific.
+
+For example, don't do this:
+```haskell 
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity tId teacher) =
+  select $ from $ \(ts `InnerJoin` cs `InnerJoin` cms) -> do
+    on $ cs ^. CourseId ==. cms  ^. CourseMembershipCourseId
+    on $ cs ^. CourseTeacherId ==. ts ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just sId ->
+          ts ^. TeacherSchoolId ==. val (Just sId)
+        Nothing ->
+          ts ^. TeacherId ==. val tId
+    pure cms
+```
+
+Do this instead:
+```haskell
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity teacherId teacher) =
+  select $ from $ \(teachers `InnerJoin` courses `InnerJoin` memberships) -> do
+    on $ courses ^. CourseId ==. memberships  ^. CourseMembershipCourseId
+    on $ courses ^. CourseTeacherId ==. teachers ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just schoolId ->
+          teachers ^. TeacherSchoolId ==. val (Just schoolId)
+        Nothing ->
+          teachers ^. TeacherId ==. val teacherId
+    pure memberships
+```
+
+One abbreviation we seem to tolerate is `num` as in `numQuestionsAnswered`. There may be others.
+
+## JavaScript
+
+JavaScript uses `camelCase` for identifiers and `TitleCase` for classes and React components. For flow types, we generally append a `T` to the end of the type, e.g. `SkillT`. Names that are being ignored should be prefixed with an underscore (e.g. `_teacher`, `_standard`). It's a common mistake (especially for Haskellers) to use a lone `_` and accidentally overwrite underscore in that scope.
+
+## Problems
+
+Serialization of string-enums to JSON and postgres is all over the place. We have examples of `snake_case`, `camelCase`, `TitleCase`, and `ALLCAPS`.
+
+If we have DB-level enums that are not shared with the frontend, it makes sense to use the postgres convention of `snake_case`. However, if we're planning to pass enums to the frontend, then we should use whatever the JavaScript convention is. JavaScript is also using pretty much everything, so it's hard to say what we should be using. I (Parks) usually use `TitleCase` for JavaScript-only enums, but only because that mimics Haskell enums.
+
+Haskell pretty much enforces a style at the language-level by requiring that the first character of each constructor be capitalized. We're using `TitleCase` everywhere except for `MathSystem` and `DomainAbbreviation` which are `ALLCAPS`.
```

Comment:
Is there a rule we should be following to determine if a new data type should be `TitleCase` or `ALLCAPS`?

Reply:
The only things that I've seen that are `ALLCAPS` are initialisms, e.g. `MD` (a domain abbreviation) or `CCSS` (Common Core State Standards). I've seen `ELA` spelled `ELA` and `Ela` though, and we pretty much always say `schoolId` not `schoolID`.

Reply:
`ALLCAPS` is also an older JavaScript style for constants. This might be the case in some places.

Reply:
YEP! We use `ALLCAPS` for some constant definition but totally not attached to it. Will be happy to change to standardize

Reply:
I use and have seen the style `ALL_CAPS` for constants in JS. Is there an alternative convention we could use?

Reply:
I think `ALL_CAPS` for constants is fine, e.g.
```jsx
const MAX_FP_SESSIONS = 2
```

I was saying I've seen `ALLCAPS` for enum values, e.g.
```jsx
type MathSystemT
  = 'CCSS'
  | 'TEKS'
  ...
const MathSystems =
  { CCSS: ('CCSS': MathSystemT)
  , TEKS: ('TEKS': MathSystemT)
  ...
  }
```

Reply:
I don't have a good heuristic for formatting short initialisms. I've seen both `HTTP` and `Http` on the backend for example.

Reply:
In general I prefer `Http` (and `Ela` over `ELA` for example) in identifiers, because they'll always start with a lowercase letter and that's easier. For example:

`elaWritingAssignment` <-> `makeElaWritingAssignmentSession` seems more consistent than:
`elaWritingAssignment` <-> `makeELAWritingAssignmentSession`

But on it's own I think I'd prefer `ELA`. Does that seem like a reasonable rule to set up? If abbreviations are on their own => capitalize, if as a part of an identifier => camelCase?